xwayland/xwm.c

```diff
@@ -65,31 +63,154 @@ static struct wlr_xwayland_surface *wlr_xwayland_surface_create(
 		wlr_log(L_ERROR, "Could not allocate wlr xwayland surface");
 		return NULL;
 	}
+
+	xcb_get_geometry_cookie_t geometry_cookie =
+		xcb_get_geometry(xwm->xcb_conn, window_id);
+
+	uint32_t values[1];
+	values[0] =
+		XCB_EVENT_MASK_FOCUS_CHANGE |
+		XCB_EVENT_MASK_PROPERTY_CHANGE;
+	xcb_change_window_attributes(xwm->xcb_conn, window_id,
+		XCB_CW_EVENT_MASK, &values);
+
+	surface->xwm = xwm;
 	surface->window_id = window_id;
 	surface->x = x;
 	surface->y = y;
 	surface->width = width;
 	surface->height = height;
 	surface->override_redirect = override_redirect;
-	wl_list_insert(&xwm->new_surfaces, &surface->link);
-	surface->state = wlr_list_create();
+	wl_list_insert(&xwm->surfaces, &surface->link);
 	wl_signal_init(&surface->events.destroy);
 	wl_signal_init(&surface->events.request_configure);
+	wl_signal_init(&surface->events.request_move);
+	wl_signal_init(&surface->events.request_resize);
+	wl_signal_init(&surface->events.request_maximize);
+	wl_signal_init(&surface->events.request_fullscreen);
+	wl_signal_init(&surface->events.map_notify);
+	wl_signal_init(&surface->events.unmap_notify);
 	wl_signal_init(&surface->events.set_class);
 	wl_signal_init(&surface->events.set_title);
 	wl_signal_init(&surface->events.set_parent);
-	wl_signal_init(&surface->events.set_state);
 	wl_signal_init(&surface->events.set_pid);
 	wl_signal_init(&surface->events.set_window_type);
+
+	xcb_get_geometry_reply_t *geometry_reply =
+		xcb_get_geometry_reply(xwm->xcb_conn, geometry_cookie, NULL);
+
+	if (geometry_reply != NULL) {
+		surface->has_alpha = geometry_reply->depth == 32;
+	}
+
+	free(geometry_reply);
+
 	return surface;
 }
 
+static void xwm_set_net_active_window(struct wlr_xwm *xwm,
+		xcb_window_t window) {
+	xcb_change_property(xwm->xcb_conn, XCB_PROP_MODE_REPLACE,
+			xwm->screen->root, xwm->atoms[_NET_ACTIVE_WINDOW],
+			xwm->atoms[WINDOW], 32, 1, &window);
+}
+
+static void xwm_send_focus_window(struct wlr_xwm *xwm,
+		struct wlr_xwayland_surface *surface) {
+	if (surface) {
+		if (surface->override_redirect) {
+			return;
+		}
+
+		xcb_client_message_event_t client_message;
+		client_message.response_type = XCB_CLIENT_MESSAGE;
+		client_message.format = 32;
+		client_message.window = surface->window_id;
+		client_message.type = xwm->atoms[WM_PROTOCOLS];
+		client_message.data.data32[0] = xwm->atoms[WM_TAKE_FOCUS];
+		client_message.data.data32[1] = XCB_TIME_CURRENT_TIME;
+
+		xcb_send_event(xwm->xcb_conn, 0, surface->window_id,
+			XCB_EVENT_MASK_SUBSTRUCTURE_REDIRECT, (char*)&client_message);
+
+		xcb_set_input_focus(xwm->xcb_conn, XCB_INPUT_FOCUS_POINTER_ROOT,
+			surface->window_id, XCB_CURRENT_TIME);
+
+		uint32_t values[1];
+		values[0] = XCB_STACK_MODE_ABOVE;
+		xcb_configure_window_checked(xwm->xcb_conn, surface->window_id,
+			XCB_CONFIG_WINDOW_STACK_MODE, values);
+	} else {
+		xcb_set_input_focus_checked(xwm->xcb_conn,
+			XCB_INPUT_FOCUS_POINTER_ROOT,
+			XCB_NONE, XCB_CURRENT_TIME);
+	}
```

Comment:
Minor issue - this would read better if you rearranged it a bit:

```c
if (!surface) {
	xcb_set_input_focus_checked(xwm->xcb_conn,
		XCB_INPUT_FOCUS_POINTER_ROOT,
		XCB_NONE, XCB_CURRENT_TIME);
	return;
}
if (surface->override_redirect) {
	return;
}
// ...
```